### ISSUE TITLE
Missing feature: Dataset templates

### ISSUE THREAD

**Issue by mih** [datalad/datalad#1462]
We have the need to (repeatedly) create datasets with a specific structure and partially uniform content. One identified case is dataset metadata (#1457) that needs to be provided in some form. However, the need for pre-configuration goes beyond that. For example, I tend to have a `code/` folder in each dataset, and I always want to have all files in this directory to be committed to git directly and not annexed. Likewise, I tend to use one and the same licence for all datasets that I produce, and my name also hasn't changed since birth.

I think we should have the ability to create datasets from a pre-configured template. Such a template could be a tarball with a pre-crafted directory structure that has pre-populated metadata files, but also `.gitattribute` files in all the right places, etc. The mechanism should be format agnostic (just files and directories). However, it might be useful to have a dedicated place in the tarball the enables further modification at the time of creation. This could be implemented as a config file at `.datalad/create_tmpl.cfg` (or whatever) inside the tarball that has a general section, and (optionally, if more tuning is needed) one section per each to be modified file. The settings in each section are key value mappings of `PLACEHOLDER = TARGETVALUE` where each target value can be overwritten giving an option for the corresponding placeholder to `create`. For example: `create mydataset --override @@LICENSE@@ CC0`, where "@@LICENSE@@" is the placeholder and "CC0" the replacement for the preconfiguration in the template.

Such a mechanism would facilitate creation of structured datasets in any form. It would also shift the paradigm (that we use in our present tests, for example) from a rather procedural creation process (first a file, then two more) to a more goal-oriented approach: "gimme a BIDS skeleton", "gimme a CRCNS dataset shell". It would also generalize the approach in #1457 to scale beyond a single metadata file. The solution for #1457 could then be used as a default "interactive" template override that enables people to be guided to all files and all setting relevant to a dataset, instead of long option chains (some multi-line values will be hard to stringify and could be easily copied into an editor).

**Comment by mih** [comment]
Any thoughts on this @yarikoptic ?

**Comment by yarikoptic** [comment]
In general I like the idea... Yet to read few more times for better diggestion

**Comment by yarikoptic** [comment]
that is what was nagging me:  such templates couldn't be "composed" together and thus would be not as flexible as they could be made if created on the fly given a set of specifications.  E.g. a particular case you brought up 
- "`code/` folder in each dataset, and I always want to have all files in this directory to be committed to git directly and not annexed".  
Although indeed could be "hard coded" within few templates, it is a generic pattern which could be useful for many other datasets and as such ideally should be just a feature you demand from a new dataset.
This reflects on old RFC #921 and #925 where we tried to discuss how we could simplify/standardize some "features" we are assigning to created siblings or new datasets.  So it feels to me that we might better come up with a list of dataset/sibling "features" (which are just a set of tags we understand) which trigger specific actions. E.g.
- `with-code` 
   - would be your usecase which would trigger action "modify .gitattributes to instruct annex to commit `code/` straight under git"
- `bids`:
   - adds/processes `with-code` feature
   - expand .gitattributes to instruct to add top-level README, *tsv etc directly to git
   - (not sure if possible actually but ...) set `.gitattributes` to instruct to add `distribution-restrictions` for `sourcedata/` content upon their commit.  Or at least, if operating on top of existing but not yet in datalad directory -- do corresponding `git-annex metadata` call
   - trigger `datalad metadata-edit --type=bids` (#1457) to establish `dataset_description.json`
   - initiate README (although may be should be done by above step)
   - I am not sure what else could be in "BIDS skeleton" (even subject IDs are nohow regimented) but that is only the level where I think some tarball template might be appropriate to consider -- just to compliment with some files layout but again as triggered by this "feature" which would do other actions
- `nodata`, `public`, `alltags`, `allbranches` (#925)
   - would tune up git annex wanted and git/config on what that dataset/sibling should contain
- `crawl`:
   - would trigger `crawl-init` in some kind of interactive mode

and may be even allow to be of the form `name=value` so we could add/pass options into corresponding actions. And then may be even allow for that simple templating from a tarball but without much on top...

E.g. examples
- `datalad create --features=with-code` - your original use case
- `datalad create --features=template=/path/mylab.tar.gz`
  - to initiate from some template, and if `metadata` type is set within its dataset config, issue `metadata-edit`  with its type
- `datalad create --features=bids:license=CC,crawl:template=openfmri`
   - in this case `license=CC` becomes an option within `datalad edit-metadata --type=bids --license=CC` call
- `datalad create-sibling --features=unique-data,public,alltags,allbranches`
   - to create a sibling to publish derivative data publicly (thus imposing restriction to not publish data with restrictions)

So altogether it feels that we could benefit from "template tarballs" but imho it should be just an additional feature within a more generic approach

**Comment by mih** [comment]
I like that in general. However, I see these as two related, but different things.

> that is what was nagging me: such templates couldn't be "composed" together

Why not? Multiple template tarballs superimposed on each other. What is most important to me is that there is a mechanism that allows me to compose datasets with a custom structure that does not need to have been envisioned by datalad authors.

**Comment by yarikoptic** [comment]
> Multiple template tarballs superimposed on each other. 

and what to happen if both tarballs provide `.gitattributes`?  I could only see approach similar to what I do in crawler:  create detached branches for both tarballs, extract, and then do 'git merge' them into a single piece.

> compose datasets with a custom structure that does not need to have been envisioned by datalad authors

you (i.e. any user) could always do that without anything special in datalad. `cp` or `rsync` your favorite template one before you call `datalad create --force` within the copy ;)  but indeed **us** providing some templates would be handy (as we envision them)

**Comment by yarikoptic** [comment]
just for the record, had to come up with this .gitattributes:
```
* annex.backend=MD5E
# top level text files -- git
* annex.largefiles=(not(mimetype=text/*))
# everything underneath subdirectories -- annex
**/* annex.largefiles=anything
```

**Comment by yarikoptic** [comment]
damn I want this feature in some reincarnation now :-/  becomes really needed for automated crawling of collections with bunch of sub-repos... e.g. if I know that they all will be BIDS datasets I would want to instruct that subdatasets should be creating following some template which in turn would provide correct values in .gitattributes one way or another

**Comment by mih** [comment]
Work towards this is happening in #1619

**Comment by mih** [comment]
FTR: @bpoldrack language to setup test repos is already capable enough to do most of this. We should use it for this purpose too, and not have yet another solution.

**Comment by bpoldrack** [comment]
Agree. Need to RF to extract that layer of interpreting that "language" and use it for templates again. Probably just needs some additional string mappings.
